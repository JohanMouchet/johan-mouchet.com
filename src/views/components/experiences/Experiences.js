// @flow

import * as React from "react";
import duration from "utils/duration";
import parse from "utils/parse";
import { Projects } from "views/components";
import "./Experiences.scss";

type Props = {
  experiences: [
    {
      company: {
        url: string,
        name: string,
      },
      location: string,
      jobTitle: string,
      startDate: Date,
      endDate: Date,
      lede: string,
      projects: Projects,
    }
  ],
};

const Experiences = ({ experiences }: Props) => {
  const today = new Date();

  return !experiences ? null : (
    <div className="c-experiences">
      {experiences.map((experience) => (
        <section
          className="c-experience"
          key={experience.company.name + " - " + experience.startDate}
        >
          <h3 className="c-experience__heading">
            <a
              href={experience.company.url && experience.company.url}
              target={experience.company.url && "_blank"}
              rel={experience.company.url && "noopener noreferrer"}
              className="c-experience__company"
            >
              {experience.company.name}
            </a>
            , {experience.location} ⁠— {experience.jobTitle}
          </h3>

          <p className="c-experience__subheading">
            <span className="c-experience__date">
              {`${experience.startDate.toLocaleString("en-GB", {
                month: "short",
              })} ${experience.startDate.getFullYear()}`}

              {experience.endDate.toDateString() === today.toDateString()
                ? " – Present"
                : [
                    (experience.endDate.getFullYear() >
                      experience.startDate.getFullYear() ||
                      experience.endDate.getMonth() >
                        experience.startDate.getMonth()) &&
                      ` –  ${experience.endDate.toLocaleString("en-GB", {
                        month: "short",
                      })}`,
                    experience.endDate.getFullYear() >
                      experience.startDate.getFullYear() &&
                      ` ${experience.endDate.getFullYear()}`,
                  ]}
            </span>
            <span className="c-experience__duration">
              {" "}
              ({duration(experience.endDate, experience.startDate, "1 mo")})
            </span>
            {experience.contractType && (
              <span className="c-experience__type">
                {" "}
                / {experience.contractType}
              </span>
            )}
          </p>

          <p className="c-experience__lede">{parse(experience.lede)}</p>

          {experience.projects && <Projects projects={experience.projects} />}
        </section>
      ))}
    </div>
  );
};

export default Experiences;
