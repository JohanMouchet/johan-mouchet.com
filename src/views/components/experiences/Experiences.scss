@import "assets/styles/settings/_settings";
@import "assets/styles/tools/_tools";

.c-experience {
    &:not(:first-child) {
      border-top: 1px solid $s-color-gray;
      padding-top: t-vertical-rhythm(1);
    }

  &__heading {
    margin-bottom: 0;
  }

  &__company {
    font-weight: 700;
  }

  &__subheading {
    margin-top: 0;
    color: $s-color-text-dimmed;
  }
}
