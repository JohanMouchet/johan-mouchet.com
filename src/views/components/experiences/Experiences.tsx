import { duration } from "@/utils/duration/duration";
import { Projects } from "@/views/components/projects/Projects";
import { PrismicRichText } from "@prismicio/react";
import clsx, { ClassValue } from "clsx";
import styles from "./Experiences.module.scss";

export const Experiences = ({
  experiences,
  className,
  ...props
}: {
  experiences: Array<{
    companyLink: {
      url: string;
    };
    companyName: string;
    location: string;
    jobTitle: string;
    startDate: Date | string;
    present: boolean;
    endDate: Date | string;
    contractType?: string;
    lede: React.ComponentProps<typeof PrismicRichText>["field"];
    projects?: React.ComponentProps<typeof Projects>;
  }>;
  className?: ClassValue;
} & React.HTMLProps<HTMLDivElement>) => {
  if (!experiences?.length) {
    return null;
  }

  return (
    <div className={clsx(styles["c-experiences"], className)} {...props}>
      {experiences.map((experience) => {
        const startDate =
          typeof experience.startDate === "string"
            ? new Date(experience.startDate)
            : experience.startDate;
        const endDate = experience.present
          ? new Date()
          : typeof experience.endDate === "string"
          ? new Date(experience.endDate)
          : experience.endDate;

        return (
          <section
            className={styles["c-experience"]}
            key={`${experience.companyName}-${startDate}`}
          >
            <h3 className={styles["c-experience__heading"]}>
              <a
                href={experience.companyLink.url && experience.companyLink.url}
                target={experience.companyLink.url && "_blank"}
                rel={experience.companyLink.url && "noopener noreferrer"}
                className={styles["c-experience__company"]}
              >
                {experience.companyName}
              </a>
              , {experience.location} ⁠— {experience.jobTitle}
            </h3>
            <p className={styles["c-experience__subheading"]}>
              <span className={styles["c-experience__date"]}>
                <time dateTime={startDate.toISOString().substring(0, 7)}>
                  {`${startDate.toLocaleString("en-GB", {
                    month: "short",
                  })} ${startDate.getFullYear()}`}
                </time>

                {experience.present ? (
                  " – Present"
                ) : (
                  <>
                    {" "}
                    –{" "}
                    <time dateTime={endDate.toISOString().substring(0, 7)}>
                      {`${endDate.toLocaleString("en-GB", {
                        month: "short",
                      })} ${endDate.getFullYear()}`}
                    </time>
                  </>
                )}
              </span>{" "}
              <span className={styles["c-experience__duration"]}>
                (
                <time dateTime={duration(startDate, endDate, "iso8601", "1M")}>
                  {duration(startDate, endDate, "short", "1 mo")}
                </time>
                )
              </span>
              {experience.contractType && (
                <span className={styles["c-experience__type"]}>
                  {" "}
                  / {experience.contractType}
                </span>
              )}
            </p>
            <div className={styles["c-experience__lede"]}>
              <PrismicRichText field={experience.lede} />
            </div>
            {experience.projects?.projects && (
              <Projects projects={experience.projects.projects} />
            )}
          </section>
        );
      })}
    </div>
  );
};
