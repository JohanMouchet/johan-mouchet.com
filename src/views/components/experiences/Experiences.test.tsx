import React from "react";
import { render } from "@testing-library/react";
import Experiences from ".";

it("Renders <Experiences /> without crashing", () => {
  expect(
    render(
      <Experiences
        experiences={[
          {
            company: {
              url: "#",
              name: "Name",
            },
            location: "Location",
            jobTitle: "Job Title",
            startDate: new Date("2018-01-01"),
            endDate: new Date("2019-01-01"),
            lede: "Lede",
            projects: [],
          },
        ]}
      />
    ).container.firstChild
  ).toMatchSnapshot();
});
