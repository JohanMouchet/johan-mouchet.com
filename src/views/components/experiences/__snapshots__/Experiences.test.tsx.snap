// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <Experiences /> without crashing 1`] = `""`;
